ls1intum/Artemis
# `Development`: Make text blocks more robust

### Checklist
#### General


#### Client

### Motivation and Context


The ID of `TextBlock`s is computed as a hash value of
- submission ID
- start index
- end index
- text

Currently, we have to remember to call `textBlock.computeId()` whenever we change any of these attributes so that the ID is always up to date. This is prone to bugs where a developer forgets this call.

Hopefully helps fix #7131.

### Description

We now always compute the text block ID on any changes to the text block attributes.

### Steps for Testing

Prerequisites:
- 1 TA
- 1 Student
- 1 Text Exercise

1. Participate in the text exercise as the student
2. Grade the exercise as the tutor. Test using existing text blocks as well as custom ones (using `Opt/Alt` in the assessment editor).
3. Save the assessment.
4. Refresh the page. Check that all feedbacks are still the same.
5. Open the assessment as the student. Check that the result is the same: Same credits, same sum, same feedback texts, same blocks, ...

#### Exam Mode Testing


Do the same in an exam. Also testing a second correction round would be ideal.

Try to reproduce #7131. This is difficult because we haven't found a reliable way to reproduce it.

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage
